lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -1591,20 +1591,10 @@ def __init__(
             if not isinstance(refresh_event, list):
                 refresh_event = [refresh_event]
 
+        self.framework.observe(events.relation_joined, self.set_scrape_job_spec)
         for ev in refresh_event:
             self.framework.observe(ev, self.set_scrape_job_spec)
 
-        # Update relation data every reinit. If instead we used event hooks then observing only
-        # relation-joined would not be sufficient:
-        # - Would need to observe leader-elected, in case there was no leader during
-        #   relation-joined.
-        # - If later related to an ingress provider, then would need to register and wait for
-        #   update-status interval to elapse before changes would apply.
-        # - The ingerss-ready custom event is currently emitted prematurely and cannot be relied
-        #   upon: https://github.com/canonical/traefik-k8s-operator/issues/78
-        # NOTE We may still end up waiting for update-status before changes are applied.
-        self.set_scrape_job_spec()
```

Comment:
I wonder why calling this every reinit is a data integrity problem.
Is it because user calls a method after construction that changes things up?

Reply:
Yes. Initially, this was due to the LXD charm calling `update_scrape_job_spec` when it saw new units/"children" with endpoints to fan in (inside the model), which got blown away on the next event of any kind.

`lookaside_jobs_callable` is a solution/workaround for the LXD charm, but in general, "rewrite the relation databag every time any event fires" won't be sustainable in the long run as our use cases increased. 

It was a workaround for the ingress problem (ironically, partially resolved by adding `update_scrape_job_spec` as we've realized that the entire state of the charm/model will not be knowable from inside a the charm constructor, before any relation events have fired), but it's not necessary anymore, and it's causing data integrity issues. "Use `lookaside_jobs_callable`" isn't a meaningful long-term workaround/fix either, unfortunately. For users who have long-running workloads with endpoints they want to manage themselves in a way which is more dynamic than `prometheus-scrape-target` but less dynamic than potentially rebuilding the list every single time, just... not overwriting it on charm init is good.